fms_extras/utils/generation.py

```diff
@@ -0,0 +1,371 @@
+import time
+from typing import Any, Callable, List, MutableMapping, Optional, Union
+
+import torch
+import torch.nn.functional as F
+
+from fms_extras.models.speculator import MLPSpeculator
+from fms_extras.utils.cache import flatten_batch, select_inflate_dim
+from fms_extras.utils.cache.paged import PagedAttentionCacheData, PagedKVCacheManager
+
+
+def speculative_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids: Union[torch.Tensor, List[torch.Tensor]],
+    speculator: MLPSpeculator,
+    kv_cache_manager: PagedKVCacheManager,
+    max_seq_len: int = 2048,
+    new_tokens: int = 256,
+    top_k: int = 5,
+    threshes=[5, 3, 2],
+    flatting=True,
+    decode_model: Optional[Union[Callable, torch.nn.Module]] = None,
+):
+    """
+    A reference implementation of speculative decoding generation.
+    Returns at least the specified number of tokens - the speculator may return a
+    few extra in the final step.
+    If input is batched, continues generating until EVERY sequence has produced AT LEAST the required number of tokens.
+    Input (and output) tokens beyond max_seq_len are simply dropped for a sliding-window approach.
+    Currently reproduces behavior of greedy decoding only.
+    Args:
+        model: A function or nn.Module that takes a batch of input_ids and
+            returns logits
+        input_ids: A length n tensor of token IDs, or list of such tensors
+        speculator: A function or nn.Module that takes a state vector and sampled token
+            and returns a set of candidate suffixes
+        kv_cache_manager: PagedKVCacheManager
+            the paged kv-cache manager to be used in generation
+        max_seq_len: the sequence length of the base model
+        new_tokens: number of tokens to generate
+        top_k: only score the top k candidates from the speculator
+        threshes: use top k predictions from each head to generate speculator candidate pool
+        verbose_dict: Optional HF tokenizer vocab dict. If provided, runs verbosely and prints
+            speculator behavior and scoring for each step
+    Returns:
+        result: List of id tensors, possibly different lengths if batching.
+        n_steps: Number of foward passes used to generate provided tokens.
+    """
+    # Construct batch(es) and initial inputs
+    if decode_model is None:
+        decode_model = model
+    bsize = len(input_ids)
+    result = input_ids  # [b] n
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids])
+    n_pads_init = [max_len - seq.size(0) for seq in input_ids]
+    n_pads = torch.Tensor(n_pads_init).to(device=input_ids[0].device, dtype=torch.int)
```

Comment:
prefer using torch.tensor() instead of torch.Tensor

Reply:
updated in latest commit